checker/src/body_visitor.rs

```diff
@@ -1498,21 +1498,21 @@ impl<'analysis, 'compilation, 'tcx, E> BodyVisitor<'analysis, 'compilation, 'tcx
         let mut no_children = true;
         if val_type == ExpressionType::NonPrimitive || is_closure {
             // First look at paths that are rooted in rpath.
-            for (path, value) in self
-                .current_environment
-                .value_map
+            for (path, value) in value_map
                 .iter()
                 .filter(|(p, _)| p.is_rooted_by(&source_path))
             {
                 check_for_early_return!(self);
                 let qualified_path = path.replace_root(&source_path, target_path.clone());
                 if move_elements {
                     trace!("moving child {:?} to {:?}", value, qualified_path);
-                    value_map = value_map.remove(path);
+                    self.current_environment.value_map =
+                        self.current_environment.value_map.remove(path);
                 } else {
                     trace!("copying child {:?} to {:?}", value, qualified_path);
                 };
-                value_map = value_map.insert(qualified_path.clone(), value.clone());
+                self.current_environment
+                    .update_value_at(qualified_path, value.clone());
                 no_children = false;
             }
         }
```

Comment:
On line 1524 & line 1537, should we operate directly on `self.current_environment.value_map` instead of on `value_map`?

Reply:
Indeed. Good catch.